### DISCUSSION TITLE
UMat is not utilising GPU with CvInvoke.Rectangle() and CvInvoke.PutText()

### DISCUSSION THREAD

**Discussion by riyasharma15** [emgucv/emgucv#966] [Q&A]
I read an image as a UMat and was trying to add a rectangle and some text on the UMat image using CvInvoke.Rectangle() and CvInvoke.PutText(). So now as UMat utilises GPU for performing operations it should have utilised GPU but when i checked the GPU utilisation using nvidia-smi -l command it was 0%. 

And when i tried to convert a coloured UMat image to a grayscale UMat image using CvInvoke.CvtColor() and tried to resize a UMat image using CvInvoke.Resize() i could see GPU being utilised.

So are there some specific operations for which UMat utilises GPU or how is it? I have no idea why it is like this. Can anyone please help me understand this?

**Comment by sn4k3** [comment]
`UMat` is a "trap" because it's an abstraction where it will use GPU just for a few supported operations, all other operations that require to run in CPU will degrade your performance with transfers. Avoid UMat at all cost unless you know exactly you are taking advtange of it (All your operations are supported in GPU). 
openCV have poor GPU implementation, not worth IMO. Only Cuda have real benefit if you manage to fit all your needs under a GpuMat.
